src/bitmap/ops.rs

```diff
@@ -264,11 +285,27 @@ impl BitAnd<&RoaringBitmap> for &RoaringBitmap {
 
     /// An `intersection` between two sets.
     fn bitand(self, rhs: &RoaringBitmap) -> RoaringBitmap {
-        if rhs.len() < self.len() {
-            BitAnd::bitand(self.clone(), rhs)
-        } else {
-            BitAnd::bitand(rhs.clone(), self)
+        let mut containers = Vec::new();
+        let mut iter_lhs = self.containers.iter().peekable();
+        let mut iter_rhs = rhs.containers.iter().peekable();
+
+        loop {
+            match (iter_lhs.peek(), iter_rhs.peek()) {
+                (None, None) => break,
+                (Some(lhs), Some(rhs)) => {
+                    if lhs.key == rhs.key {
+                        let (lhs, rhs) = iter_lhs.next().zip(iter_rhs.next()).unwrap();
+                        let container = BitAnd::bitand(lhs, rhs);
+                        if container.len != 0 {
+                            containers.push(container);
+                        }
+                    }
+                }
+                _otherwise => (),
```

Comment:
is this not an infinite loop if you try to do the intersection of an empty set and a non emtpy set?

Reply:
Wow, thank you very much, that's indeed a potential infinite loop!